File: gallery/src/main/java/com/qkopy/gallery/ui/imagepicker/RecyclerViewManager.java
/*
 * Created by @ajithvgiri on 4/12/18 8:14 PM
 * Copyright (c) 2018 . All rights reserved.
 * Last modified 4/12/18 7:05 PM
 */

package com.qkopy.gallery.ui.imagepicker;

import android.content.Context;
import android.content.res.Configuration;
import android.os.Parcelable;
import android.widget.Toast;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.qkopy.gallery.R;
import com.qkopy.gallery.adapter.FolderPickerAdapter;
import com.qkopy.gallery.adapter.ImagePickerAdapter;
import com.qkopy.gallery.listener.OnBackAction;
import com.qkopy.gallery.listener.OnFolderClickListener;
import com.qkopy.gallery.listener.OnImageClickListener;
import com.qkopy.gallery.listener.OnImageSelectionListener;
import com.qkopy.gallery.model.Config;
import com.qkopy.gallery.model.Folder;
import com.qkopy.gallery.model.Image;
import com.qkopy.gallery.widget.GridSpacingItemDecoration;

import java.util.ArrayList;
import java.util.List;


public class RecyclerViewManager {

    private Context context;
    private RecyclerView recyclerView;
    private Config config;

    private GridLayoutManager layoutManager;
    private GridSpacingItemDecoration itemOffsetDecoration;

    private ImagePickerAdapter imageAdapter;
    private FolderPickerAdapter folderAdapter;

    private int imageColumns;
    private int folderColumns;


    private Parcelable foldersState;
    private String title;
    private boolean isShowingFolder;


    public RecyclerViewManager(RecyclerView recyclerView, Config config, int orientation) {
        this.recyclerView = recyclerView;
        this.config = config;
        context = recyclerView.getContext();
        changeOrientation(orientation);
        isShowingFolder = config.isFolderMode();
    }

    public void setupAdapters(OnImageClickListener imageClickListener, final OnFolderClickListener folderClickListener) {
        ArrayList<Image> selectedImages = null;
        if (config.isMultipleMode() && !config.getSelectedImages().isEmpty()) {
            selectedImages = config.getSelectedImages();
        }

        imageAdapter = new ImagePickerAdapter(context, selectedImages, imageClickListener);
        folderAdapter = new FolderPickerAdapter(context, new OnFolderClickListener() {
            @Override
            public void onFolderClick(Folder folder) {
                foldersState = recyclerView.getLayoutManager().onSaveInstanceState();
                folderClickListener.onFolderClick(folder);
            }
        });
    }

    /**
     * Set item size, column size base on the screen orientation
     */
    public void changeOrientation(int orientation) {
        imageColumns = orientation == Configuration.ORIENTATION_PORTRAIT ? 3 : 5;
        folderColumns = orientation == Configuration.ORIENTATION_PORTRAIT ? 2 : 4;

        int columns = isShowingFolder ? folderColumns : imageColumns;
        layoutManager = new GridLayoutManager(context, columns);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setHasFixedSize(true);
        setItemDecoration(columns);
    }

    private void setItemDecoration(int columns) {
        if (itemOffsetDecoration != null) {
            recyclerView.removeItemDecoration(itemOffsetDecoration);
        }
        itemOffsetDecoration = new GridSpacingItemDecoration(columns,
                context.getResources().getDimensionPixelSize(R.dimen.item_padding),
                false
        );
        recyclerView.addItemDecoration(itemOffsetDecoration);
        layoutManager.setSpanCount(columns);
    }


    public void setOnImageSelectionListener(OnImageSelectionListener imageSelectionListener) {
        checkAdapterIsInitialized();
        imageAdapter.setOnImageSelectionListener(imageSelectionListener);
    }

    public List<Image> getSelectedImages() {
        checkAdapterIsInitialized();
        return imageAdapter.getSelectedImages();
    }

    public void addSelectedImages(List<Image> images) {
        imageAdapter.addSelected(images);
    }

    private void checkAdapterIsInitialized() {
        if (imageAdapter == null) {
            throw new IllegalStateException("Must call setupAdapters first!");
        }
    }

    public boolean selectImage() {
        if (config.isMultipleMode()) {
            if (imageAdapter.getSelectedImages().size() >= config.getMaxSize()) {
                String message = String.format(config.getLimitMessage(), config.getMaxSize());
                Toast.makeText(context, message, Toast.LENGTH_SHORT).show();
                return false;
            }
        } else {
            if (imageAdapter.getItemCount() > 0) {
                imageAdapter.removeAllSelected();
            }
        }
        return true;
    }

    public void handleBack(OnBackAction action) {
        if (config.isFolderMode() && !isShowingFolder) {
            setFolderAdapter(null);
            action.onBackToFolder();
            return;
        }
        action.onFinishImagePicker();
    }

    public void setImageAdapter(List<Image> images, String title) {
        imageAdapter.setData(images);
        setItemDecoration(imageColumns);
        recyclerView.setAdapter(imageAdapter);
        this.title = title;
        isShowingFolder = false;
    }

    public void setFolderAdapter(List<Folder> folders) {
        folderAdapter.setData(folders);
        setItemDecoration(folderColumns);
        recyclerView.setAdapter(folderAdapter);
        isShowingFolder = true;

        if (foldersState != null) {
            layoutManager.setSpanCount(folderColumns);
            recyclerView.getLayoutManager().onRestoreInstanceState(foldersState);
        }
    }

    public String getTitle() {
        if (isShowingFolder) {
            return config.getFolderTitle();
        } else if (config.isFolderMode()) {
            return title;
        } else {
            return config.getImageTitle();
        }
    }

    public boolean isShowDoneButton() {
        return config.isMultipleMode() && (config.isAlwaysShowDoneButton() || imageAdapter.getSelectedImages().size() > 0);
    }

    public int selectedMediaCount() {
        return imageAdapter.getSelectedImages().size();
    }
}

File: gallery/src/main/java/com/qkopy/gallery/ui/imagepicker/RecyclerViewManager.kt
package com.qkopy.gallery.ui.imagepicker

import android.content.Context
import android.content.res.Configuration
import android.os.Parcelable
import android.widget.Toast
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.qkopy.gallery.R
import com.qkopy.gallery.adapter.FolderPickerAdapter
import com.qkopy.gallery.adapter.ImagePickerAdapter
import com.qkopy.gallery.listener.OnBackAction
import com.qkopy.gallery.listener.OnFolderClickListener
import com.qkopy.gallery.listener.OnImageClickListener
import com.qkopy.gallery.listener.OnImageSelectionListener
import com.qkopy.gallery.model.Config
import com.qkopy.gallery.model.Folder
import com.qkopy.gallery.model.Image
import com.qkopy.gallery.widget.GridSpacingItemDecoration

class RecyclerViewManager(
    private val recyclerView: RecyclerView,
    private val config: Config,
    orientation: Int
) {
    private val context: Context
    private var layoutManager: GridLayoutManager? = null
    private var itemOffsetDecoration: GridSpacingItemDecoration? = null
    private var imageAdapter: ImagePickerAdapter? = null
    private var folderAdapter: FolderPickerAdapter? = null
    private var imageColumns = 0
    private var folderColumns = 0
    private var foldersState: Parcelable? = null
    private var title: String? = null
    private var isShowingFolder: Boolean
    fun setupAdapters(
        imageClickListener: OnImageClickListener?,
        folderClickListener: OnFolderClickListener
    ) {
        var selectedImages: ArrayList<Image>? = null
        if (config.isMultipleMode && !config.selectedImages!!.isEmpty()) {
            selectedImages = config.selectedImages
        }
        imageAdapter = ImagePickerAdapter(context, selectedImages, imageClickListener!!)
        folderAdapter = FolderPickerAdapter(context, object : OnFolderClickListener {
            override fun onFolderClick(folder: Folder) {
                foldersState = recyclerView.layoutManager!!.onSaveInstanceState()
                folderClickListener.onFolderClick(folder)
            }
        })
    }

    /**
     * Set item size, column size base on the screen orientation
     */
    fun changeOrientation(orientation: Int) {
        imageColumns =
            if (orientation == Configuration.ORIENTATION_PORTRAIT) 3 else 5
        folderColumns =
            if (orientation == Configuration.ORIENTATION_PORTRAIT) 2 else 4
        val columns = if (isShowingFolder) folderColumns else imageColumns
        layoutManager = GridLayoutManager(context, columns)
        recyclerView.layoutManager = layoutManager
        recyclerView.setHasFixedSize(true)
        setItemDecoration(columns)
    }

    private fun setItemDecoration(columns: Int) {
        if (itemOffsetDecoration != null) {
            recyclerView.removeItemDecoration(itemOffsetDecoration!!)
        }
        itemOffsetDecoration = GridSpacingItemDecoration(
            columns,
            context.resources.getDimensionPixelSize(R.dimen.item_padding),
            false
        )
        recyclerView.addItemDecoration(itemOffsetDecoration!!)
        layoutManager!!.spanCount = columns
    }

    fun setOnImageSelectionListener(imageSelectionListener: OnImageSelectionListener?) {
        checkAdapterIsInitialized()
        imageAdapter!!.setOnImageSelectionListener(imageSelectionListener)
    }

    val selectedImages: List<Image>
        get() {
            checkAdapterIsInitialized()
            return imageAdapter!!.getSelectedImages()
        }

    fun addSelectedImages(images: List<Image>?) {
        imageAdapter!!.addSelected(images)
    }

    private fun checkAdapterIsInitialized() {
        checkNotNull(imageAdapter) { "Must call setupAdapters first!" }
    }

    fun selectImage(): Boolean {
        if (config.isMultipleMode) {
            if (imageAdapter!!.getSelectedImages().size >= config.maxSize) {
                val message =
                    String.format(config.limitMessage!!, config.maxSize)
                Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
                return false
            }
        } else {
            if (imageAdapter!!.itemCount > 0) {
                imageAdapter!!.removeAllSelected()
            }
        }
        return true
    }

    fun handleBack(action: OnBackAction) {
        if (config.isFolderMode && !isShowingFolder) {
            setFolderAdapter(null)
            action.onBackToFolder()
            return
        }
        action.onFinishImagePicker()
    }

    fun setImageAdapter(
        images: List<Image>?,
        title: String?
    ) {
        imageAdapter!!.setData(images)
        setItemDecoration(imageColumns)
        recyclerView.setAdapter(imageAdapter)
        this.title = title
        isShowingFolder = false
    }

    fun setFolderAdapter(folders: List<Folder>?) {
        folderAdapter!!.setData(folders)
        setItemDecoration(folderColumns)
        recyclerView.setAdapter(folderAdapter)
        isShowingFolder = true
        if (foldersState != null) {
            layoutManager!!.spanCount = folderColumns
            recyclerView.layoutManager!!.onRestoreInstanceState(foldersState)
        }
    }

    fun getTitle(): String? {
        return if (isShowingFolder) {
            config.folderTitle
        } else if (config.isFolderMode) {
            title
        } else {
            config.imageTitle
        }
    }

    val isShowDoneButton: Boolean
        get() = config.isMultipleMode && (config.isAlwaysShowDoneButton || imageAdapter!!.getSelectedImages().size > 0)

    fun selectedMediaCount(): Int {
        return imageAdapter!!.getSelectedImages().size
    }

    init {
        context = recyclerView.context
        changeOrientation(orientation)
        isShowingFolder = config.isFolderMode
    }
}
